zebra-state/src/service/check/issuance.rs

```diff
@@ -59,9 +72,13 @@ pub fn valid_burns_and_issuance(
                 .apply_change(change)
                 .ok_or(ValidateContextError::InvalidIssuance)?;
 
-            issued_assets
-                .insert(asset_base, updated_asset_state)
-                .expect("transactions must have only one burn item per asset base");
+            // FIXME: Is it correct to do nothing if the issued_assets aready has asset_base? Now it'd be
+            // replaced with updated_asset_state in this case (where the duplicated value is added to
+            // the supply). Block may have two burn records for the same asset but in different
+            // transactions - it's allowed, that's why the check has been removed. On the other
+            // hand, there needs to be a check that denies duplicated burn records for the same
+            // asset inside the same transaction.
```

Comment:
It does need to check that there are no duplicate burns in a single transaction _somewhere_, but that could be validated during deserialization or semantic validation, and this should update the asset state correctly otherwise.

I think we should update the `Burn` type to be `Burn(HashMap<AssetBase, NoteValue>)` and return an error during deserialization if any asset base is burned twice in the same transaction.

Reply:
That's a good idea to use a `HashMap` to ensure the uniqueness of `AssetBase` at the type level. In this case, the uniqueness would be validated in the deserialization function. However, I see two downsides:

1. Using a `HashMap` makes the `burn.rs` module code more bloated, as, for example, we can't directly take advantage of the automatic vector serialization provided by `ZcashSerialize` and `ZcashDeserialize` and have to implement it manually. But this isn't a major issue - I attempted to implement it today.

2. A more serious problem is that `HashMap` does not preserve the order of items. If we deserialize the transaction and then serialize it back or calculate the hash, the result would differ because the order of the burn items would change (most likely different, to be precise). To resolve this issue, we would need to add a new requirement to the specification that burn items must be sorted in the transaction by the bytes of `AssetBase`, for example.

In my opinion, we could continue using a `Vec` for now but perform uniqueness validation within/after the deserializer. And we already have a burn validation function in the `orchard` crate. I will add this check.